### PR TITLE
enable rand(::Union{AbstractSet,Associative})

### PR DESCRIPTION
**Pull request by rfourquet**
This was previously implemented only for concrete types from base (i.e. `Set`, `IntSet`, `Dict`). 
~~WIP as I don't manage to put some method definitions of `GenericSet` and `GenericDict` in test.jl (so I put them in test/random.jl in the meantime), someone has an idea?~~ cf. #22288.

### CODE REVIEW

File: test/random.jl [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
 
+# should be in base/test.jl, but does not compile...
+
+for (G, A) in ((GenericSet, AbstractSet),
+               (GenericDict, Associative))
+   @eval begin
+      Base.convert(::Type{$G}, s::$A) = $G(s)
+       Base.done(s::$G, state) = done(s.s, state)
+      Base.next(s::$G, state) = next(s.s, state)
+  end
```
Comment (tkelman) [inline test/random.jl]: indent's off here

**Comment by rfourquet** [comment]
@stevengj , @ararslan , thanks for your help, I moved to this PR the warning about linear complexity in the general case (as it is solved now for strings in #22224).

**Review by ararslan — APPROVED** [review]

File: base/random.jl [hunk 2]
```diff
@@ -257,13 +257,21 @@ globalRNG() = GLOBAL_RNG
...
 
 `S` defaults to [`Float64`](@ref).
 
+!!! note
+    The complexity of `rand(rng, s::Union{Associative,AbstractSet})`
+    is linear in the length of `s`, unless an optimized method (with
+    constant complexity) is available (which is the case for `Dict`,
+    `Set` and `IntSet`). For more than a few calls, use `rand(rng,
+    collect(s))` instead (or `rand(rng, Dict(s))`, or `rand(rng,
+    Set(s))`).
```
Comment (tkelman) [inline base/random.jl]: the frequent parentheticals are a bit distracting here, the "(which is the case for ...)" is the only one that wouldn't work fine as just a normal part of the sentence

Comment (ararslan) [inline base/random.jl]: Actually even that would be fine as part of the sentence, IMO.
> The complexity of `rand(rng, s::Union{Associative,AbstractSet})` is linear in the length of `s`, unless an optimized method with constant complexity is available, which is the case for `Dict`, `Set`, and `IntSet`. For more than a few calls, use `rand(rng, collect(s))` instead, or `rand(rng, Dict(s))` or `rand(rng, Set(s))` as appropriate.

Comment (rfourquet) [inline base/random.jl]: Ok thank you both, I updated. I also put this note after the examples, which seem more important to me in most of the cases.

**Comment by ararslan** [comment]
Looks like this needs a rebase but then should be good to go.

File: base/test.jl [hunk 3]
```diff
@@ -11,6 +11,8 @@ test set that throws on the first failure. Users can choose to wrap
 their tests in (possibly nested) test sets that will store results
 and summarize them at the end of the test set with `@testset`.
 """
+:Test # cf. #22288
```
Comment (tkelman) [inline base/test.jl]: guess this commit should be squashed then, on merge if not before

Comment (rfourquet) [inline base/test.jl]: I would say all commits can be squashed on merge, and I can squash before if prefered.

**Pull request merged**